internal/runtime/client/client.go

```diff
@@ -116,33 +116,32 @@ type hookClient struct {
 	hook   catalog.Hook
 }
 
-func (h *hookClient) CallAll(ctx context.Context, request, response runtime.Object) error {
+// TODO: Using an interface that is in hooksv1alpha1 means now the client has a tight dependency with the api version.
+// Same goes for the Call function while using the GenericExtensionResponse and the `ResponseStatusSuccess`.
+func (h *hookClient) CallAll(ctx context.Context, request runtime.Object, response runtimehooksv1.AggregatableResponse) (runtimehooksv1.HookResponseSummary, error) {
 	gvh, err := h.client.catalog.GroupVersionHook(h.hook)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	registrations, err := h.client.registry.List(gvh)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	responses := []runtime.Object{}
+	responses := []*runtimehooksv1.ExtensionResponse{}
 	for _, registration := range registrations {
-		response, err := h.client.catalog.NewResponse(gvh)
+		tmpResponse, err := h.client.catalog.NewResponse(gvh)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		// Follow-up: Call all extensions irrespective of error. Don't short-circuit on an error.
-		if err := h.Call(ctx, registration.Name, request, response); err != nil {
-			return err
-		}
-		responses = append(responses, response)
+		err = h.Call(ctx, registration.Name, request, tmpResponse)
+		responses = append(responses, &runtimehooksv1.ExtensionResponse{
+			ExtensionName: registration.Name,
+			Response:      tmpResponse,
+			Error:         errors.Wrapf(err, "extension %v failed", registration.Name),
+			FailurePolicy: registration.FailurePolicy,
+		})
 	}
-	h.aggregateResponses(responses, response)
-	return nil
-}
-
-func (h *hookClient) aggregateResponses(list []runtime.Object, into runtime.Object) {
-	// TODO; panic("implement a response aggregation mechanism")
+	return response.Aggregate(responses)
 }
 
 func (h *hookClient) Call(ctx context.Context, name string, request, response runtime.Object) error {
```

Comment:
Here are my findings:
In Kubernetes admission webhooks, the effective response is either `Allowed` or `Denied`. The `Status` field in the `AdmissionResponse` of k8s is only used to provide additional information when the response is `Denied`. The `Status` field in k8s admission response does not serve the same purpose as the `Status` field in our Runtime Hooks as per our RuntimeSDK design doc.


In Kubernetes, there are 2 kinds of errors that the webook code deals with:
1) Error when calling the webhook (example: network work because the webhook is unreachable) - ErrCallingWebhook.
2) Internal error caused in core for other reasons - InternalError.

There is no equivalent of an error where the webhook successfully returns a response (therefore now a network error) but the response.Status == Failure.

The Failure Policy is only used to deal with first kind of errors, ErrCallingWebhook. If the failurePolicy is `Ignore` and we get an ErrCallingWebhook the webhook handler acts as if the webhook never existed and continues with calling the next admission webhook. If the failurePolicy is `Fail`, it fails immediately and the next webhook is not called.

For InternalError, we error out immediately and we dont call the next webook.

**Thoughts on how we should be handling `FailurePolicy` when working with RuntimeHooks:**
We have 3 kinds of errors:
1) Error when calling the webhook (example: network work because the webhook is unreachable)
2) Error received as a response from the webhook (Response received. Response.Status == Failure)
3) Internal errors

For the purposes of error handling I think we should deal with the first and the second kind of errors the same way as both are errors from the webhook (and external system) and should be handled similar to like how Kubernetes handles ErrCallingWebhook (affected by FailurePolicy).

Reply:
FailurePolicy dictates how unrecognized errors from the external system.

An unrecognized error is any error that does not allow us to get a response.

FailurePolicy == `Ignore`  instruct the system to continue when we have an unrecognized error (Example: when calling a runtime extension.)

What to do with `Response.Status == failure` is dictated by the application. (FailurePolicy does not have an affect on this).